builtin/providers/aws/config.go

```diff
@@ -316,3 +326,34 @@ func (c *Config) ValidateAccountId(iamconn *iam.IAM) error {
 
 	return nil
 }
+
+// This function is responsible for reading credentials from the
+// environment in the case that they're not explicitly specified
+// in the Terraform configuration.
+func getCreds(key, secret, token string) *awsCredentials.Credentials {
+	// build a chain provider, lazy-evaulated by aws-sdk
+	providers := []awsCredentials.Provider{
+		&awsCredentials.StaticProvider{Value: awsCredentials.Value{
+			AccessKeyID:     key,
+			SecretAccessKey: secret,
+			SessionToken:    token,
+		}},
+		&awsCredentials.EnvProvider{},
+		&awsCredentials.SharedCredentialsProvider{},
+	}
+
+	// We only look in the EC2 metadata API if we can connect
+	// to the metadata service within a reasonable amount of time
+	metadataURL := os.Getenv("AWS_METADATA_URL")
+	if metadataURL == "" {
+		metadataURL = "169.254.169.254:80"
+	}
+	c := http.Client{
+		Timeout: 100 * time.Millisecond,
+	}
+	_, err := c.Get(metadataURL)
+	if err == nil {
```

Comment:
Let's also log in both cases - if we decide we _are_ on an EC2 instance _or_ if we decide we are not.

We've had a bunch of weird client behavior around this so I think it helps to be extra verbose here.